### DISCUSSION TITLE
what's the proper way to use non-trivial web workers across domains?

### DISCUSSION THREAD

**Discussion by dr-matt** [webpack/webpack#17007] [Q&A]
Apologies if this has been answered before, but I have only been able to find pieces of solutions here and there.

I have a web application that contains a standalone react component bundled using webpack. This component spawns web workers, and these workers load and execute WebAssembly code, which is all included in the bundle. Everything works fine if the application and the react component are hosted on the same server.

I now have a need to serve the react component from a CDN, splitting it from the main application's host. The worker was originally created using the syntax that allows webpack to statically identify and bundle the worker script:

```
new Worker(new URL('../workers/my.worker.ts', import.meta.url))
```

Suppose the main app is served from app.com and the component is served from cdn.com. The bundle on cdn.com is accessible and its entry point loads fine from app.com, but when `new Worker()` is called, it causes this error:

```
DOMException: Failed to construct 'Worker': Script at 'https://cdn.com/component/component.647.chunk.js' cannot be accessed from origin 'https://app.com'
```

I could not find a way to make this work across domains, even with `Content-Security-Policy: worker-src` headers. So I implemented [this workaround](https://github.com/webpack/webpack/discussions/14648#discussioncomment-1589272) from @piotr-oles, which uses `importScripts` inside a blob URL and aliases `Worker` to preserve the special webpack syntax. With this change, the worker is accessible and loads partially, but then I see errors like this:

```
DOMException: Failed to execute 'importScripts' on 'WorkerGlobalScope': The script at 'https://cdn.com/component/component.647.chunk.js' failed to load.
```

Tracking down the cause of this reveals that the worker script itself is trying to load another bundled chunk, since the worker code contains its own imports, but it is requesting the wrong URL: it's trying to load `blob:https://app.com/component.402.chunk.js` when it should be trying to load `https://cdn.com/component/component.402.chunk.js`. It seems that webpack's internal `importScripts` call is using the wrong prefix - that of the temporary blob URL from the workaround above - to try to load other bundled resources from the worker.

So, I'm stumped with several questions.
1. Is it impossible to just create the worker directly using the `new Worker()` syntax, even if I have control of both app.com and cdn.com?
2. If so, is the referenced blob URL method the recommended workaround, even if the worker script calls `import` / needs to load other chunks?
3. Perhaps I have the wrong approach altogether - what is missing here? Do I need to set `publicPath` and/or the new `workerPublicPath` in the webpack config, or use different http headers, or change something else in the config or code...?

Any guidance is appreciated!

**Comment by alexander-akait** [comment]
The most easy solution - proxy them, so you will not have `origin` problems, it is not hard to setup

**Reply by dr-matt** [reply]
Thanks @alexander-akait. Would you kindly elaborate / point to an example? Is that the recommended best practice here, or can it work along the lines of the above approach?

**Reply by alexander-akait** [reply]
You can setup it for dev server (or for your server) using https://webpack.js.org/configuration/dev-server/#devserverproxy, I can't say your configuration, because I don't know your structure of a project, but it is simple to setup

**Reply by dr-matt** [reply]
Many thanks. For reference, I'm using apache to serve the main application on app.com, and using amazon cloudfront to serve the react component and workers on cdn.com. If I understand correctly, you're suggesting setting up apache to proxy requests for the react component, so to the browser those resources appear to be served from app.com (same origin), but in reality the proxy is pulling them from cdn.com on the server side. Is that what you have in mind? If so, that approach seems counter to the main purpose of using the cdn, which is to offload / distribute traffic, be more responsive, and support a flexible deployment architecture.

Are there other options / suggestions for getting this to work?

**Reply by alexander-akait** [reply]
> If so, that approach seems counter to the main purpose of using the cdn, which is to offload / distribute traffic, be more responsive, and support a flexible deployment architecture.

I suggested this for use in development mode because it's easier to set up and works great, not for production

For production usage you need to setup right headers, of course you can create worker from blob, but it is not a good solution

**Comment by dr-matt** [comment]
I see, thank you. Fortunately in dev mode, everything works fine - there are no cross origin issues since all content is served from localhost, and workers can be created in the normal way without blobs.

I agree that creating a worker from a blob is not the ideal solution. It seems hacky and causes other problems downstream, like the one described here. But I haven't been able to find a working solution for production; even with CORS headers and CSP `worker-src` header, I still see the original `failed to construct Worker` error mentioned above. My understanding from reading other discussions was that using blobs is the only way around this, but if there's a better method, I'd much prefer that.

**Reply by alexander-akait** [reply]
In theory you can try to use `importScripts`, but if your worker is in ESM, it will not work

**Comment by piotr-oles** [comment]
Can you use absolute publicPath or do you need relative one?
The issue is when you try to import another script using relative path. It's because blob worker has `blob://...` origin. You can use absolute paths or overwrite `importScript()` function like here: https://github.com/webpack/webpack/discussions/14648#discussioncomment-2559969

**Reply by dr-matt** [reply]
@piotr-oles I do need relative `publicPath` as this component is used in different directory locations for different apps. And you're exactly right about the new `blob://` origin. 

Thank you for the link on replacing `importScripts`. That code didn't work for me as written when I tried it a while back, but I was eventually able to make it work with some even more hacky modifications. Unfortunately it leads to other downstream problems, as my particular worker also needs (and fails) to load emscripten-compiled webassembly modules included in the webpack bundle. I will try continuing down that path, but it doesn't feel like a stable long-term solution - is there really no cleaner way to serve components like this from a different host?

**Reply by dr-matt** [reply]
Update: it looks like the webassembly module load is failing because the underlying javascript glue is using `fetch` to get the wasm resource, and it's trying to fetch from origin `blob:...` just like `importScripts` does. So if I apply the same function replacement trick to redefine `self.fetch`, I can get everything to load.

A remaining issue is that the component is typically served from a subdirectory under the web root, and that subdirectory is omitted from the url in some (but not all)  `importScripts` and `fetch` calls. I don't know yet how to resolve this.

**Comment by dr-matt** [comment]
With a combination of the clever solutions from @piotr-oles and @OmarCastro in [this thread](https://github.com/webpack/webpack/discussions/14648#discussioncomment-2559969), and a few additional modifications, I think everything is working now. For reference, this required:

- creating a wrapper that aliases `Worker` to a custom class which loads the worker script from a blob via `importScripts` while still retaining webpack syntax compatibility;
- replacing the native `self.importScripts` function in the worker context with a custom version that corrects the origin and base path of the input url before import (this allows the worker to import bundled scripts without the incorrect `blob` origin);
- replacing the worker's native `self.fetch` function in a similar way (this allows the worker to load bundled wasm modules);
- using `__webpack_public_path__` rather than `window.location.href` in these replacement functions to properly set both the correct origin and correct directory path.

It's certainly possible that there are remaining corner cases not handled by this solution, and it still feels brittle, but it's working well enough now to move onward. Perhaps this can be addressed within webpack sometime in the future.
